File: frontend/src/api/posts.ts
import axiosClient from './axiosClient'

export interface PostContent {
  body: string
  tags: string[]
  cover_image_url?: string
  description?: string
}

export interface Post {
  id: number
  title: string
  slug: string
  author_id: number
  author_username?: string
  mongo_id: string
  visibility: string
  content_type: string
  likes_count: number
  claps_count: number
  cover_image_url?: string
  created_at: string
  updated_at?: string
}

export interface PostWithContent extends Post {
  content: PostContent
}

export interface PostCreate {
  title: string
  slug: string
  visibility: string
  content_type: string
  content: PostContent
}

export interface PostUpdate {
  title?: string
  slug?: string
  visibility?: string
  content?: PostContent
  content_type?: string
}

export interface PostListResponse {
  posts: Post[]
  total: number
  page: number
  page_size: number
}

export const postsApi = {
  createPost: async (data: PostCreate): Promise<Post> => {
    const response = await axiosClient.post('/api/v1/posts', data)
    return response.data
  },

  getPosts: async (page = 1, pageSize = 20, sortBy = 'latest', search?: string, contentType?: string): Promise<PostListResponse> => {
    const params: any = { page, page_size: pageSize, sort_by: sortBy }
    if (search) params.search = search
    if (contentType) params.content_type = contentType
    const response = await axiosClient.get('/api/v1/posts', { params })
    return response.data
  },

  getPost: async (postId: number): Promise<PostWithContent> => {
    const response = await axiosClient.get(`/api/v1/posts/${postId}`)
    return response.data
  },

  getPostBySlug: async (slug: string): Promise<PostWithContent> => {
    const response = await axiosClient.get(`/api/v1/posts/slug/${slug}`)
    return response.data
  },

  likePost: async (postId: number): Promise<Post> => {
    const response = await axiosClient.post(`/api/v1/posts/${postId}/like`)
    return response.data
  },

  clapPost: async (postId: number): Promise<Post> => {
    const response = await axiosClient.post(`/api/v1/posts/${postId}/clap`)
    return response.data
  },

  getPostEngagement: async (postId: number): Promise<{
    is_liked: boolean
    has_clapped: boolean
    likes_count: number
    claps_count: number
  }> => {
    const response = await axiosClient.get(`/api/v1/posts/${postId}/engagement`)
    return response.data
  },

  getMyPosts: async (includeDrafts = false): Promise<Post[]> => {
    const response = await axiosClient.get('/api/v1/posts/me', {
      params: { include_drafts: includeDrafts },
    })
    return response.data
  },

  updatePost: async (postId: number, data: PostUpdate): Promise<Post> => {
    const response = await axiosClient.put(`/api/v1/posts/${postId}`, data)
    return response.data
  },

  deletePost: async (postId: number): Promise<void> => {
    await axiosClient.delete(`/api/v1/posts/${postId}`)
  },

  searchPosts: async (params: {
    page?: number
    page_size?: number
    sort_by?: string
    search?: string
    author?: string
    content_type?: string
    genre?: string
    start_date?: string
    end_date?: string
    full_text?: boolean
  }): Promise<PostListResponse> => {
    const cleanParams: any = {}
    if (params.page) cleanParams.page = params.page
    if (params.page_size) cleanParams.page_size = params.page_size
    if (params.sort_by) cleanParams.sort_by = params.sort_by
    if (params.search) cleanParams.search = params.search
    if (params.author) cleanParams.author = params.author
    if (params.content_type) cleanParams.content_type = params.content_type
    if (params.genre) cleanParams.genre = params.genre
    if (params.start_date) cleanParams.start_date = params.start_date
    if (params.end_date) cleanParams.end_date = params.end_date
    if (params.full_text !== undefined) cleanParams.full_text = params.full_text

    const response = await axiosClient.get('/api/v1/posts/search', { params: cleanParams })
    return response.data
  },
}

File: backend/app/routes/posts.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import Optional
from sqlalchemy.orm import Session
from app.database.postgres import get_db
from app.schemas.post_schema import PostCreate, PostResponse, PostUpdate, PostWithContent, PostListResponse
from app.services.post_service import (
    create_post, get_post, get_post_by_slug, get_post_content,
    update_post, delete_post, get_public_posts, get_user_posts
)
from app.utils.dependencies import get_current_user
from app.models.user import User

router = APIRouter(prefix="/posts", tags=["posts"])

@router.post("", response_model=PostResponse, status_code=status.HTTP_201_CREATED)
async def create_new_post(
    post_data: PostCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a new post"""
    # Check if slug already exists
    existing_post = get_post_by_slug(db, post_data.slug)
    if existing_post:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Slug already exists"
        )
    
    post = await create_post(db, post_data, current_user.id)  # Add await here
    return post 

@router.get("", response_model=PostListResponse)
async def get_posts(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: str = Query("latest", regex="^(latest|most_appreciated)$"),
    search: Optional[str] = Query(None, description="Search query for post titles"),
    content_type: Optional[str] = Query(None, description="Filter by content type (article, poetry, book)"),
    db: Session = Depends(get_db)
):
    """Get public posts with pagination, search, and filtering"""
    skip = (page - 1) * page_size
    posts, total = get_public_posts(db, skip=skip, limit=page_size, sort_by=sort_by, search=search, content_type=content_type)
    
    # Add author usernames and cover images to posts
    from app.models.user import User
    from app.services.post_service import get_post_content
    posts_with_author = []
    for post in posts:
        author = db.query(User).filter(User.id == post.author_id).first()
        post_dict = post.__dict__.copy()
        post_dict['author_username'] = author.username if author else None
        
        # Get cover image from MongoDB
        try:
            content = await get_post_content(post.mongo_id)
            if content:
                post_dict['cover_image_url'] = content.cover_image_url
        except:
            post_dict['cover_image_url'] = None
        
        posts_with_author.append(PostResponse(**post_dict))
    
    return PostListResponse(
        posts=posts_with_author,
        total=total,
        page=page,
        page_size=page_size
    )


@router.get("/me", response_model=list[PostResponse])
async def get_my_posts(
    include_drafts: bool = Query(False),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get current user's posts"""
    posts = get_user_posts(db, current_user.id, include_drafts=include_drafts)
    return posts


@router.get("/{post_id}", response_model=PostWithContent)
async def get_post_by_id(
    post_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get post by ID with content"""
    post = get_post(db, post_id)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    # If post is a draft, only the author or admin can access it
    if post.visibility == "draft":
        if post.author_id != current_user.id and not current_user.is_admin:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized to access this draft"
            )
    
    content = await get_post_content(post.mongo_id)
    if not content:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Post content not found"
        )
    
    # Add author username
    from app.models.user import User
    author = db.query(User).filter(User.id == post.author_id).first()
    post_dict = post.__dict__.copy()
    post_dict['author_username'] = author.username if author else None
    
    return PostWithContent(
        **post_dict,
        content=content
    )


@router.get("/slug/{slug}", response_model=PostWithContent)
async def get_post_by_slug_route(
    slug: str,
    db: Session = Depends(get_db)
):
    """Get post by slug with content"""
    post = get_post_by_slug(db, slug)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    if post.visibility == "draft":
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    content = await get_post_content(post.mongo_id)
    if not content:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Post content not found"
        )
    
    # Add author username
    from app.models.user import User
    author = db.query(User).filter(User.id == post.author_id).first()
    post_dict = post.__dict__.copy()
    post_dict['author_username'] = author.username if author else None
    
    return PostWithContent(
        **post_dict,
        content=content
    )


@router.put("/{post_id}", response_model=PostResponse)
async def update_post_by_id(
    post_id: int,
    post_data: PostUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update post"""
    post = get_post(db, post_id)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    if post.author_id != current_user.id and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to update this post"
        )
    
    update_data = post_data.dict(exclude_unset=True)
    updated_post = await update_post(db, post, update_data)
    return updated_post


@router.delete("/{post_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_post_by_id(
    post_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete post"""
    post = get_post(db, post_id)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    if post.author_id != current_user.id and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to delete this post"
        )
    
    await delete_post(db, post)
    return None


@router.post("/{post_id}/like", response_model=PostResponse)
async def like_post(
    post_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Like a post (toggle)"""
    post = get_post(db, post_id)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    if post.visibility == "draft":
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    # Check if user already liked
    if current_user in post.liked_by:
        # Unlike
        post.liked_by.remove(current_user)
        post.likes_count = max(0, post.likes_count - 1)
    else:
        # Like
        post.liked_by.append(current_user)
        post.likes_count += 1
    
    db.commit()
    db.refresh(post)
    return post


@router.post("/{post_id}/clap", response_model=PostResponse)
async def clap_post(
    post_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Clap for a post (adds to count, can clap multiple times)"""
    post = get_post(db, post_id)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    if post.visibility == "draft":
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    # Add clap (users can clap multiple times)
    post.claps_count += 1
    if current_user not in post.clapped_by:
        post.clapped_by.append(current_user)
    
    db.commit()
    db.refresh(post)
    return post


@router.get("/{post_id}/engagement")
async def get_post_engagement(
    post_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get user's engagement status for a post"""
    post = get_post(db, post_id)
    if not post:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Post not found"
        )
    
    is_liked = current_user in post.liked_by if current_user else False
    has_clapped = current_user in post.clapped_by if current_user else False
    
    return {
        "is_liked": is_liked,
        "has_clapped": has_clapped,
        "likes_count": post.likes_count,
        "claps_count": post.claps_count
    }

